### PR TITLE
Refactor PotentialTemplates as a lazy loading json module(Closes #333)

### LINKED ISSUE #333: How to create Potential Templates? User Guide as well as discussion
We dont have much documentation for contributing potentials to gmso. This should be updated. 

Other contributions should probably be highlighted too.

### PR DESCRIPTION
**Pull request by umesh-timalsina**
This PR addresses #333, Refactors PotentialTemplate classes by consolidating them into a singleton instance that has a modified `__getitem__`. The information about `PotentialTemplates` is now saved as importable JSON files, which is loaded lazily on demand from the code base. This also adds documentation on how to contribute a `PotentialTemplate` as well as refactors few other files to incorporate the changed PotentialTemplates API.

### CODE REVIEW

**Comment by bdice** [comment]
I think you may need to specify those JSON files in setup.py's `package_data` argument and `MANIFEST.in` to ensure they're included. By default they won't be included.
https://setuptools.readthedocs.io/en/latest/setuptools.html#including-data-files
https://packaging.python.org/guides/using-manifest-in/

See signac-flow for an example. We have to make sure to package templates (which are .sh files). https://github.com/glotzerlab/signac-flow/blob/master/MANIFEST.in

Test this by making a source distribution or wheel build of the package, installing it, and then making sure it imports correctly from another directory (not the package root directory).

**Comment by umesh-timalsina** [comment]
@bdice. Thanks I will add them to `setup.py`.

File: gmso/lib/jsons/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  "name": "OPLSTorsionPotential",
+  "expression": "0.5 * k0 + 0.5 * k1 * (1 + cos(phi)) + 0.5 * k2 * (1 - cos(2*phi)) + 0.5 * k3 * (1 + cos(3*phi)) + 0.5 * k4 * (1 - cos(4*phi))",
+  "independent_variables": "phi"
+}
+```
+If you want to add a template named `ABCTemplate`. Create a file called
+`ABCTemplate.json` in this directory. Using the json reference above, 
+name the template `ABCTemplate` and change the `expression` and `independent_variables`.
+
+If there are more than one  `independent_variables` in your template, simply separate them by a comma(,).
```
Comment (justinGilmer) [inline gmso/lib/jsons/README.md]:
```suggestion
If there are more than one  `independent_variables` in your template, simply separate them with a comma(,).
```

small grammar nit.

File: gmso/lib/jsons/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+2. **Use a python script**:
+Let's say you want to create a new PotentialTemplate. There's a static method 
+in `PotentialTemplates` that let's you add a new one.
+
+Use the following script to generate potential template(in your dev branch):
+
+**Warning: Do not use this script on conda/pip installed version or master branch**
```
Comment (justinGilmer) [inline gmso/lib/jsons/README.md]: Can we go into a bit more detail why that is the case? I think this is the case because the template json file will be added to the site-packages of the installed `gmso`, correct?

File: gmso/lib/potential_templates.py [hunk 3]
```diff
@@ -1,158 +1,130 @@
+import glob
+import os
+import json
+
 from gmso.core.potential import Potential
+from gmso.utils.singleton import Singleton
+from gmso.exceptions import GMSOError
 
+POTENTIAL_JSONS = glob.glob(os.path.join(os.path.dirname(__file__), 'jsons', '*.json'))
```
Comment (justinGilmer) [inline gmso/lib/potential_templates.py]: I know the `os` module is the normal go to, but would it make sense to use [the `pathlib` module here?](https://docs.python.org/3/library/pathlib.html)
Ive been using them for a while now and they are quite nice, and i think they provide the same utility we use here, just a more higher-level api.

File: gmso/lib/potential_templates.py [hunk 4]
```diff
@@ -1,158 +1,130 @@
...
+            exists_package = False
+        if item not in self._user_ref_dict:
+            exists_user = False
+
+        if exists_user:
+            return self._load_from_json(item, ref='user')
+        if exists_package:
+            return self._load_from_json(item, ref='package')
+
+        raise KeyError(f'Potential Template {item} not found in user as well as package templates'
```
Comment (justinGilmer) [inline gmso/lib/potential_templates.py]:
```suggestion
        raise KeyError(f'Potential Template {item} not found in user or the default package templates'
```

Comment (justinGilmer) [reply gmso/lib/potential_templates.py]: Maybe something along those lines

File: gmso/tests/test_potential_templates.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        assert mie_potential.name == 'MiePotential'
+        assert mie_potential.expression == sympy.sympify(
+            '(n/(n-m)) * (n/m)**(m/(n-m)) * epsilon * ((sigma/r)**n - (sigma/r)**m)')
+        assert mie_potential.independent_variables == {sympy.sympify('r')}
+
+    def test_opls_torsion_potential(self, templates):
+        opls_torsion_potential = templates['OPLSTorsionPotential']
+        assert opls_torsion_potential.name == 'OPLSTorsionPotential'
+        assert opls_torsion_potential.expression == sympy.sympify('0.5 * k0 + 0.5 * k1 * (1 + cos(phi)) + '
+                                                                  '0.5 * k2 * (1 - cos(2*phi)) +'
```
Comment (justinGilmer) [inline gmso/tests/test_potential_templates.py]: could we make these lines a bit more similar either with the `+` at the end of each str, or at the beginning?

File: gmso/tests/test_potential_templates.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                                                   {
+                                                       'name': 'TestTemplate',
+                                                       'expression': 'a+b+c',
+                                                       'independent_variables': 'c,a'
+                                                   },
+                                                   update=True,
+                                                   user_template=False
+                                                   )
+
+        assert os.path.exists(os.path.join(JSON_DIR, 'TestTemplate.json'))
```
Comment (justinGilmer) [inline gmso/tests/test_potential_templates.py]: Is there a way to ensure that when running these tests, the test json files are cleaned up?

File: MANIFEST.in [hunk 7]
```diff
@@ -0,0 +1 @@
+include gmso/lib/jsons/*.json
```
Comment (mattwthompson) [inline MANIFEST.in]: Surely there  is more we need in this file? Unrelated to these changes

File: gmso/lib/jsons/.gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+TestTemplate.json
```
Comment (mattwthompson) [inline gmso/lib/jsons/.gitignore]: Is there a reason to have this file here and not at the root of the repo?

File: gmso/utils/singleton.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class Singleton(object):
+    _inst = None
+
+    def __new__(cls):
+        if cls._inst is None:
+            cls._inst = super(Singleton, cls).__new__(cls)
+        return cls._inst
```
Comment (mattwthompson) [inline gmso/utils/singleton.py]: Can you explain this to me in plain English? It's a class that returns a super of itself?

File: gmso/tests/test_potential_templates.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+from gmso.lib.potential_templates import PotentialTemplates, JSON_DIR
+from gmso.exceptions import GMSOError
+from gmso.tests.base_test import BaseTest
+
+
+class TestPotentialTemplates(BaseTest):
+    @pytest.fixture
+    def templates(self):
+        return PotentialTemplates()
```
Comment (mattwthompson) [inline gmso/tests/test_potential_templates.py]: Not a big deal but it might be useful to have this fixture in `BaseTest`

File: gmso/tests/test_potential_templates.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+    def test_singleton_behavior(self, templates):
+        assert id(templates) == id(PotentialTemplates())
+        assert id(PotentialTemplates()) == id(PotentialTemplates())
+
+    def test_lennard_jones_potential(self, templates):
```
Comment (mattwthompson) [inline gmso/tests/test_potential_templates.py]: This and the next 10 or so tests are effectively SIMD, would it be useful to wrap them up into a single test?

File: gmso/tests/test_potential_templates.py (hunk 6 above)
Comment (mattwthompson) [reply gmso/tests/test_potential_templates.py]: Would this be achieved with a tempdir fixture?

File: gmso/lib/potential_templates.py [hunk 12]
```diff
@@ -1,158 +1,130 @@
...
-                 template=True):
 
-        super(PotentialTemplate, self).__init__(
-            name=name,
-            expression=expression,
-            independent_variables=independent_variables,
-            template=template,
-        )
+def _create_hidden_dir(new_dir=None):
+    gmso_dir = os.path.join(os.path.expanduser('~'), '.gmso')
```
Comment (mattwthompson) [inline gmso/lib/potential_templates.py]: I feel like this will work for 90% of users, but I also wonder if there is already a best-practices way of doing this? It seems like something that is nearly ubiquitous, for which there may be some community standard.

File: gmso/lib/potential_templates.py [hunk 13]
```diff
@@ -1,158 +1,130 @@
...
-                 independent_variables={'phi'}):
-
-        super(PotentialTemplate, self).__init__(
-                name=name,
-                expression=expression,
-                independent_variables=independent_variables,
-                template=True
-        )
 
+class PotentialTemplates(Singleton):
```
Comment (mattwthompson) [inline gmso/lib/potential_templates.py]: I'm not a fan of this name, being one character off of what it is made up on

File: gmso/lib/potential_templates.py [hunk 14]
```diff
@@ -1,158 +1,130 @@
...
+    """A singleton collection of all the potential templates"""
+
+    def __init__(self):
+        try:
+            self.json_refs
+        except AttributeError:
+            self.json_refs = POTENTIAL_JSONS
+            self.user_jsons = []
+            potential_names = [(os.path.split(filename)[-1]).replace('.json', '') for filename in POTENTIAL_JSONS]
```
Comment (mattwthompson) [inline gmso/lib/potential_templates.py]: I think this is a case in which `pathlib` can be a little more elegant than chopping off the extension and splitting the string up

File: gmso/utils/singleton.py (hunk 9 above)
Comment (umesh-timalsina) [reply gmso/utils/singleton.py]: A singleton class is analogous to a singleton set. As a singleton set can have only one member, a singleton class can have only one object that is created. Regardless of number of calls to the constructor. If you look at the fixture:

https://github.com/umesh-timalsina/gmso/blob/4237bb184d3bc76f66a990e74e046a3d7a428ca3/gmso/tests/test_singleton.py#L5-L8

All the objects created will have same id.

File: gmso/lib/jsons/README.md (hunk 2 above)
Comment (umesh-timalsina) [reply gmso/lib/jsons/README.md]: Yeah, I think we should rollback this change, as user defined templates are not super helpful according to our previous discussions.

Comment (umesh-timalsina) [reply gmso/lib/jsons/README.md]: The reason it happens is because we are not taking the directory as inputs but assuming that it will always be `gmso/gmso/lib/jsons`

File: gmso/lib/potential_templates.py (hunk 3 above)
Comment (umesh-timalsina) [reply gmso/lib/potential_templates.py]: Agreed.

File: MANIFEST.in (hunk 7 above)
Comment (umesh-timalsina) [reply MANIFEST.in]: I think so. Should be added in another PR.

File: gmso/lib/jsons/.gitignore (hunk 8 above)
Comment (umesh-timalsina) [reply gmso/lib/jsons/.gitignore]: It would not be necessary to add .gitignore if we don't provide a python script to save a file which I think we shouldn't. This was added because there a test that adds the file without deleting it. So, as @justinGilmer suggested, I will remove the file(from the test after completion).

File: gmso/lib/potential_templates.py (hunk 14 above)
Comment (umesh-timalsina) [reply gmso/lib/potential_templates.py]: Agreed.

File: gmso/tests/test_potential_templates.py (hunk 11 above)
Comment (umesh-timalsina) [reply gmso/tests/test_potential_templates.py]: This is done because of aiding future contributors with a contributing guideline. Other Implementations could trim the test, I agree.

File: gmso/tests/test_potential_templates.py (hunk 6 above)
Comment (rsdefever) [reply gmso/tests/test_potential_templates.py]: If there isn't a fixture this is a code snippet I've been using to contain files I'm creating in tests:

```python
import os
import tempfile
import contextlib
import shutil

@contextlib.contextmanager
def temporary_cd():
    tmp_dir = tempfile.mkdtemp()
    prev_dir = os.getcwd()
    os.chdir(os.path.abspath(tmp_dir))
    try:
        yield
    finally:
        os.chdir(prev_dir)
        shutil.rmtree(tmp_dir)
```

Comment (bdice) [reply gmso/tests/test_potential_templates.py]: I think you can do a bit better by using the temporary directory as a context manager. I also think you can skip the `abspath` this way:
```python
import os
import tempfile
import contextlib

@contextlib.contextmanager
def temporary_cd():
    prev_dir = os.getcwd()
    with tempfile.TemporaryDirectory() as tmp_dir:
        try:
            os.chdir(tmp_dir)
            yield
        finally:
            os.chdir(prev_dir)
```

File: MANIFEST.in (hunk 7 above)
Comment (bdice) [reply MANIFEST.in]: I suggest adding the license file, if nothing else. This ensures that distributions of your package include the license, which is required by conda-forge.

File: gmso/utils/singleton.py (hunk 9 above)
Comment (bdice) [reply gmso/utils/singleton.py]: You might add a docstring explaining what a singleton class is (a class of which only one instance can exist). See here for some docstring examples that you might adapt. https://stackoverflow.com/a/7346105/2505544

File: gmso/lib/potential_templates.py (hunk 13 above)
Comment (umesh-timalsina) [reply gmso/lib/potential_templates.py]: @mattwthompson Suggestion needed for this.

Comment (uppittu11) [reply gmso/lib/potential_templates.py]: In Monday's meeting I had suggested `PotentialTemplateLibrary` or `PotentialLibrary`

Comment (umesh-timalsina) [reply gmso/lib/potential_templates.py]: How about `TemplatePotentialsLibrary`?

File: gmso/lib/potential_templates.py [hunk 15]
```diff
@@ -1,158 +1,69 @@
...
 
-        super(PotentialTemplate, self).__init__(
-            name=name,
-            expression=expression,
-            independent_variables=independent_variables,
-            template=True,
-        )
 
-class HarmonicAnglePotential(PotentialTemplate):
+class PotentialTemplate(Potential):
```
Comment (uppittu11) [inline gmso/lib/potential_templates.py]: Another thing we discussed in the Monday meeting is that this class almost exactly the same as `Potential`, with the exception of the `template=True` argument supplied to the `super.__init__`. Having an entirely separate class just for changing a single flag seems unnecessarily complicated.

Comment (umesh-timalsina) [reply gmso/lib/potential_templates.py]: I lean towards refactoring it in a different PR. I can do it here if that is desirable.

**Comment by justinGilmer** [comment]
re-running azp to see if that fixes the errors, after we resolve that, im good to merge this,

**Comment by umesh-timalsina** [comment]
The merge by dfebf5b should have been done by `ours` strategy. All the tests are screwed up now.

File: gmso/lib/jsons/README.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+A potential template object requires the following attributes:
+
+1. `name`: Name of the potential template
+2. `expression`: expression for the potential template
+3. `independent_variables`: Independent variables for the template 
+
+
+## For Developers
+To add/contribute a new `PotentialTemplate` there are the following ways:
```
Comment (daico007) [inline gmso/lib/jsons/README.md:14]: Minor comment, since there is only one way to contribute to the Potential Template now, I think we should fix this sentence. May be just have it as `To add/contribute a new 'PotentialTemplate':`

Comment (justinGilmer) [reply gmso/lib/jsons/README.md:14]: agreed, but I think we can move that to a later pr. I think it is now in a merge-able state.
